cupy/random/generator.py

```diff
@@ -344,9 +344,72 @@ def shuffle(self, a):
         if a.ndim == 0:
             raise TypeError('An array whose ndim is 0 is not supported')
 
-        sample = cupy.zeros((len(a)), dtype=numpy.int32)
-        curand.generate(self._generator, sample.data.ptr, sample.size)
-        a[:] = a[cupy.argsort(sample)]
+        a[:] = a[self.permutation(len(a))]
+
+    def permutation(self, num):
+        """Returns a permuted range."""
+        if not isinstance(num, int):
+            raise TypeError('The data type of argument "num" must be integer')
+
+        sample = cupy.empty((num), dtype=numpy.int32)
+        curand.generate(self._generator, sample.data.ptr, num)
+        if num <= 128 or num > 32 * 1024 * 1024:
+            array = cupy.argsort(sample)
+        else:
+            array = cupy.arange(num, dtype=numpy.int32)
+            # apply sort of cache blocking
+            block_size = 1 * 1024 * 1024
+            while num / block_size > 12:
+                block_size *= 2
+            for j_start in range(0, num, block_size):
+                j_end = j_start + block_size
+                _cupy_permutation()(array, sample, j_start, j_end, size=num)
+        return array
+
+
+def _cupy_permutation():
+    return core.ElementwiseKernel(
+        'raw int32 _array, raw int32 _sample, int32 j_start, int32 _j_end',
+        '',
+        '''
+            int j = (sample[i] & 0x7fffffff) % num;
+            if (j == i || j < j_start || j >= j_end) continue;
+            int j_offset = ((2*j - i + num) % (num - 1)) + 1;
+            int loops = 256;
+            bool do_next = true;
+            while (do_next && loops > 0) {
+                // try to swap the contents of array[i] and array[j]
+                if (i != j) {
+                    int val_j = atomicExch(&array[j], invalid);
+                    if (val_j != invalid) {
+                        int val_i = atomicExch(&array[i], invalid);
+                        if (val_i != invalid) {
+                            array[i] = val_j;
+                            array[j] = val_i;
+                            do_next = false;
+                            // done
+                        }
+                        else {
+                            // restore array[j]
+                            array[j] = val_j;
+                        }
+                    }
+                }
+                // update j
+                j = (j + j_offset) % num;
+                loops--;
+            }
+        ''',
+        'cupy_permutation',
+        loop_prep='''
+            int *array = reinterpret_cast<int*>(&_array[0]);
+            const int *sample = reinterpret_cast<const int*>(&_sample[0]);
+            const int invalid = -1;
+            const int num = _ind.size();
+            int j_end = _j_end;
+            if (j_end > num) j_end = num;
```

Comment:
How about using `min`?

Reply:
Do you support integer `min` in cupy kernel?

Reply:
Sorry, `min` is available in my environment, but not sure if it's guaranteed to exist.
So please leave as it is.